kitchen/modules/weights/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+		StoredValue get(stored_value): u32;
+	}
+}
+
+// A "scale" to weigh transactions. This scale can be used with any transactions that take a
+// single argument of type u32. The ultimate weight of the transaction is the / product of the
+// transaction parameter and the field of this struct.
+pub struct Linear(u32);
+
+// The actual weight calculation happens in the
+// impl WeighData block
+impl WeighData<(&u32,)> for Linear {
+	fn weigh_data(&self, (x,): (&u32,)) -> Weight {
+
+		// Use saturation so that an extremely large parameter value
+		// Does not cause overflow.
+		x.saturating_mul(self.0)
+	}
+}
+
+// Any struct that is used to weigh data must also implement ClassifyDispatchInfo. Here we calssify
+// the transaction as Normal (as opposed to operational.)
+impl<T> ClassifyDispatch<T> for Linear {
+	fn classify_dispatch(&self, _: T) -> DispatchClass {
+		// Classify all calls as Normal (which is the default)
+		Default::default()
+	}
+}
+
+// Another scale to weight transactions. This one is more complex. / It computes weight according
+// to the formula a*x^2 + b*y + c where / a, b, and c are fields in the struct, and x and y are
+// transaction / parameters.
+pub struct Quadratic(u32, u32, u32);
+
+impl WeighData<(&u32, &u32)> for Quadratic {
+	fn weigh_data(&self, (x, y): (&u32, &u32)) -> Weight {
+
+		let ax2 = x.saturating_mul(*x).saturating_mul(self.0);
+		let by = y.saturating_mul(self.1);
+		let c = self.2;
+
+		ax2.saturating_add(by).saturating_add(c)
+	}
+}
+
+impl<T> ClassifyDispatch<T> for Quadratic {
+	fn classify_dispatch(&self, _: T) -> DispatchClass {
+		// Classify all calls as Normal (which is the default)
+		Default::default()
+	}
+}
+
+// A final scale to weight transactions. This one weighs / transactions where the first parameter
+// is bool. If / the bool is true, then the weight is linear in the / second parameter. Otherwise
+// the weight is constant.
+pub struct Conditional(u32);
+
+impl WeighData<(&bool, &u32)> for Conditional {
+	fn weigh_data(&self, (switch, val): (&bool, &u32)) -> Weight {
+
+		if *switch {
+			val.saturating_mul(self.0)
+		}
+		else {
+			self.0
+		}
+	}
+}
+
+impl<T> ClassifyDispatch<T> for Conditional {
+	fn classify_dispatch(&self, _: T) -> DispatchClass {
+		// Classify all calls as Normal (which is the default)
+		Default::default()
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+		// Store value does not loop at all so a fixed weight is appropriate. Fixed weights can
+		// be assigned using types available in the Substrate framework. No custom coding is
+		// necessary.
+		#[weight = SimpleDispatchInfo::FixedNormal(100)]
+		fn store_value(_origin, entry: u32) -> Result {
+
+			StoredValue::put(entry);
+
+			Ok(())
+		}
+
+		// add_n sets the storage value n times, so it should cost n times as much as
+		// store_value. Because it performs both a read and a write, the multiplier is set to 200
```

Comment:
are all reads and writes the same cost? So `get(), put()` costs twice as much as `get()` for all storage items? Doesn't this imply that the cost of `get()` and `put()` are equal?

Also, this method performs `get(), put()` the same number of times as the input `n` (in the for loop). Does this not effect the required multiplier? I would expect the overall costs to be `n * cost_of_put() + n * cost_of_get()`. I'm unsure whether the assigned weight attribute reflects this expectation.

Reply:
More like it _assumes_ they are equal. I don't know whether that's a good assumption or not and I don't know what a better estimate is. The important part is that it grows linearly with n. So whatever it costs to `get` and `put` in a particular runtime, only the constant needs to be adjusted.